File: DSA - Data Structures And Algorithm/Java/Postorder Traversal on Binary Tree (Non recursive).java
// In this program we will perform postorder traversal on binary tree (non recursively) or we can say using stack

import java.util.Stack;
import java.util.Scanner;

class Node {
    int val;
    Node left, right;

    Node(int data) {
        val = data;
        left = null;
        right = null;
    }
}

public class BinaryTree {

    static Node buildTree() {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter data : ");
        int data = scanner.nextInt();

        if (data == -1) {
            return null;
        }

        Node root = new Node(data);

        System.out.println("Enter data in left of " + data);
        root.left = buildTree();

        System.out.println("Enter data in right of " + data);
        root.right = buildTree();

        return root;
    }

    static void postOrderTraversal(Node root) {
        if (root == null) {
            return;
        }

        Stack<Node> stack = new Stack<>();
        Node lastVisited = null;
        Node current = root;

        while (current != null || !stack.isEmpty()) {
            while (current != null) {
                stack.push(current);
                current = current.left;
            }

            current = stack.peek();

            if (current.right != null && current.right != lastVisited) {
                current = current.right;
            } else {
                System.out.print(current.val + " ");
                lastVisited = current;
                stack.pop();
                current = null;
            }
        }
    }

    public static void main(String[] args) {
        Node root = buildTree();
        System.out.println("Post-order Traversal:");
        postOrderTraversal(root);
    }
}

File: Data Structures And Algorithm/C/Postorder Traversal on Binary Tree (Non recursive).c
// In this program we will perform postorder traversal on binary tree (non recursively) or we can say using stack

#include <stdio.h>
#include <stdlib.h>

struct Node {
    int val;
    struct Node* left;
    struct Node* right;
};

struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->val = data;
    node->left = NULL;
    node->right = NULL;
    return node;
}

struct Node* BuildTree() {
    int data;
    printf("Enter data: ");
    scanf("%d", &data);

    if (data == -1) return NULL;

    struct Node* root = newNode(data);

    printf("Enter data in left of %d\n", data);
    root->left = BuildTree();

    printf("Enter data in right of %d\n", data);
    root->right = BuildTree();

    return root;
}

void Postorder(struct Node* root) {
    if (root == NULL) return;

    struct Node* stack[100]; 
    struct Node* lastVisited = NULL;
    struct Node* current = root;
    int top = -1;

    while (current != NULL || top != -1) {
        while (current != NULL) {
            stack[++top] = current;  
            current = current->left; 
        }

        current = stack[top];

        if (current->right != NULL && current->right != lastVisited) {
            current = current->right;
        } else {
            printf("%d ", current->val);
            lastVisited = current;
            current = NULL;
            top--;
        }
    }
}

int main() {
    struct Node* root = NULL;
    root = BuildTree();
    printf("Postorder Traversal: ");
    Postorder(root);
    return 0;
}
